### PR TITLE
[ValueTracking] Add support for overflow detection functions is `isKnownNonZero`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for overflow detection functions is `isKnownNonZero`; NFC**
- **[ValueTracking] Split `isNonZero(mul)` logic to a helper; NFC**
- **[ValueTracking] Add support for overflow detection functions is `isKnownNonZero`**

### CODE REVIEW

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/InstSimplify/known-non-zero.ll [hunk 1]
```diff
@@ -166,3 +165,293 @@ A:
 B:
   ret i1 0
 }
+
+declare void @use.i1(i1)
+declare void @use.i8(i1)
+define i1 @extract_value_uadd(i8 %xx, i8 %yy) {
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/test/Transforms/InstSimplify/known-non-zero.ll:171]: I think you do understand correctly. I think basically all of our extra logic for `isNonZeroAdd` is really dup logic of proving `nuw`.

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/test/Transforms/InstSimplify/known-non-zero.ll:171]: Fixed (made it a vec test w/ partial nuw where we can prove for the demandedelts).

File: llvm/test/Transforms/InstCombine/known-bits.ll [hunk 2]
```diff
@@ -630,5 +630,374 @@ define i8 @or_ne_bits_must_be_unset2_fail(i8 %x, i8 %y) {
...
+; CHECK-NEXT:    ret i1 false
+;
+  %x = add nuw i8 %xx, 1
+  %y = add nuw i8 %yy, 1
+  %x_lemma = icmp ult i8 %x, 128
+  %y_lemma = icmp ult i8 %y, 128
+  call void @llvm.assume(i1 %x_lemma)
+  call void @llvm.assume(i1 %y_lemma)
+
+  %add_sov = call { i8, i1} @llvm.sadd.with.overflow(i8 %x, i8 %y)
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/known-bits.ll:792]:
```suggestion
  %add_sov = call { i8, i1 } @llvm.sadd.with.overflow(i8 %x, i8 %y)
```
Here and elsewhere.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -2786,6 +2791,32 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
     // handled in isKnownNonZero.
     return false;
   }
+  case Instruction::ExtractValue: {
+    const WithOverflowInst *WO;
+    if (match(I, m_ExtractValue<0>(m_WithOverflowInst(WO)))) {
+      switch (WO->getIntrinsicID()) {
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2797]: Could slightly simplify this by matching on getBinaryOp().

**Review by nikic — APPROVED** [review]
LGTM
